### ISSUE TITLE
[Feature Request] Data parallel training support

### ISSUE THREAD

**Issue by dmakoviichuk-tt** [tenstorrent/tt-metal#13835]
**Is your feature request related to a problem? Please describe.**
I need to use CCL to send neural network weights from one device to another without using host.
Also we need to have all_reduce support for the gradients.
**Describe the solution you'd like**
CCL ops:
broadcast and all_reduce/reduce

**Additional context**
It will unlock LLM training on multiple devices with really good performance.

**Comment by SeanNijjar** [comment]
This overlaps with two separate feature requests:
1) All-reduce op support 
2) Send-receive op support

There are issues to track progress for these. 

If there is a training project dashboard I would like to add those issues to that dashboard. Will keep this issue open for now.

**Comment by prajaramanTT** [comment]
@dmakoviichuk-tt Can you add priority to this ticket ?

**Comment by SeanNijjar** [comment]
Hey @dmakoviichuk-tt Can I please get more details about what are your priority configurations/use cases for each of these? This will let us prioritize the more important configurations (since for example, a mesh broadcast is different from a bidirectional ring broadcast is different from parallel line broadcasts).


## All Reduce:
- For all-reduce I plan on providing a "fused" optimal implementation (today the "optimized" implementation is a composite of reduce scatter followed by all-gather). This will be line topology initially to target the TG Llama perf use case (we launch n parallel all-reduce operations along rows or columns of the TG).
- Future work (lower priority currently): Ring all-reduce
- Not yet planned: mesh all-reduce

## Broadcast:
We can start work on this soon but which use cases are most valued?
- Mesh broadcast
- line/ring broadcast
- Alternatively, if you are targeting device categories instead, then that will also be useful (e.g. ring and mesh on n300 don't really make sense)

## Reduce
Is this higher or lower priority than broadcast? Doing the same exercise as with broadcast will be helpful.

**Comment by fahmedchanTT** [comment]
@ntarafdar This looks to be a legacy ccls op issue. Can you help with what the next step would be here?

**Comment by ntarafdar** [comment]
- all_reduce support we have through an all-gather and local reduce
- broadcast and reduce support is coming soon as part of the WH CCL generality push. I'd expect early September. if this is more critical we can push it up.
- If workarounds are required we can use the a2a_broadcast op today to replicate any CCL we need

**Comment by cmaryanTT** [comment]
@ntarafdar Is this solved now with existing CCL generality?

**Comment by ntarafdar** [comment]
ya can close
